MediaSource::m_private is accessed before initialization

szewai opened the pull request:
#### 92b08efceed7b8d85a725179fe4b95cb89fef698
<pre>
MediaSource::m_private is accessed before initialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=282365">https://bugs.webkit.org/show_bug.cgi?id=282365</a>
<a href="https://rdar.apple.com/138955983">rdar://138955983</a>

Reviewed by Chris Dumez.

When MediaSource::m_client is initialized, MediaSourceClientImpl is constructed. The MediaSourceClientImpl constructor
reads parent.m_private, which is MediaSource::m_private; but MediaSource::m_private is undefined as it is initailized
after MediaSource::m_client. To fix this, ensure m_private is initialized before m_client in MediaSource.

* Source/WebCore/Modules/mediasource/MediaSource.h:

Canonical link: <a href="https://commits.webkit.org/285977@main">https://commits.webkit.org/285977@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd10a7e79183a9726cf5099263786c5842bbd48e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74345 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53774 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27156 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78725 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25583 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76462 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62907 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1559 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58439 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16766 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77412 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48602 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63953 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38848 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45591 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21448 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23916 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66995 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21795 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80243 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1662 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/957 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66723 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1810 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63971 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66007 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9945 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/8104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11482 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1626 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4414 "Failed to checkout and rebase branch from PR 35974") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1655 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1643 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1662 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/mediasource/MediaSource.cpp:
```diff
@@ -230,6 +230,7 @@ MediaSource::MediaSource(ScriptExecutionContext& context, MediaSourceInit&& opti
 #if !RELEASE_LOG_DISABLED
     , m_logger(logger(context))
 #endif
+    , m_private(nullptr)
```
Explicitly setting a RefPtr to nullptr is not right.

szewai replied:
Removed.

cdumez commented on Source/WebCore/Modules/mediasource/MediaSource.h:
```diff
@@ -251,8 +251,8 @@ class MediaSource
     std::atomic<uint64_t> m_associatedRegistryCount { 0 };
-    Ref<MediaSourceClientImpl> m_client;
     RefPtr<MediaSourcePrivate> m_private;
+    Ref<MediaSourceClientImpl> m_client;
```
Could we add a comment stating that this data member MUST be after m_private? Otherwise, it is very easy to regress.

szewai committed:
MediaSource::m_private is accessed before initialization
https://bugs.webkit.org/show_bug.cgi?id=282365
rdar://138955983

Reviewed by Chris Dumez.

When MediaSource::m_client is initialized, MediaSourceClientImpl is constructed. The MediaSourceClientImpl constructor
reads parent.m_private, which is MediaSource::m_private; but MediaSource::m_private is undefined as it is initailized
after MediaSource::m_client. To fix this, ensure m_private is initialized before m_client in MediaSource.

* Source/WebCore/Modules/mediasource/MediaSource.h:

Canonical link: https://commits.webkit.org/285977@main
